### DISCUSSION TITLE
Citation Tracking Back All the Way to the Original Text Chunk?

### DISCUSSION THREAD

**Discussion by CHENGLING-QIU** [microsoft/graphrag#800] [Q&A]
Hi team,
Question: Is there an existing solution to trace back the citation all the way to the original text?

More background:
I would like to trace back from the citation, e.g. Repots(ID1, ID2), Entities(ID1, ID2) all the way to the original text chunks; so that I can easily make sure if the generation is actual relevant. This is critical because even just looking at the traced report or entity level, I would have to worry about if these contents are hallucinated. 

I didn't find discussion on that after some research (the most relevant ask seems to be [issue_729](https://github.com/microsoft/graphrag/issues/729)

To achieve that, I am looking into the response object from the search. Looking into `result.context_data` & `result.context_text`, there are actually the contents for these context.

However:
1. It seems all the items (definitely more than these in the response) are returned. Ideally, I really just want these got cited. I am not sure if these cited types and corresponding IDs are included from the search result somehow?
e.g.
`result.context_data["reports"]` returns all the reports seemingly, not just these cited in this particular search
Context data:
```
      id  ... rank
0    284  ...  9.0
1    385  ...  8.5
2    271  ...  8.5
3    381  ...  8.5
4    322  ...  8.5
..   ...  ...  ...
343  318  ...  8.5
344  169  ...  8.5
345  159  ...  7.5
346   33  ...  8.5
347  440  ...  8.5
```

2. Again, it seems there is no way to directly trace back to the original text chunk before any GraphRAG processing. Even the `context_text `contains the internal content that has the continuous citation to other level. 

**It would be nice if the functionality tracing back to the original text is already built in, and if so, what's the best to retrieve that?
If not, it will still be great if you could give me some suggestions how to implement given a search response.**

Thank you!

**Comment by KylinMountain** [comment]
i have implemented please check out my fork of graph tag.

![IMG_9020](https://github.com/user-attachments/assets/aa30dca3-3662-4fe9-9c7f-b4c203d1c01d)
![IMG_9021](https://github.com/user-attachments/assets/e81667d9-402d-46cf-adb8-0eba455291f0)

**Reply by CHENGLING-QIU** [reply]
> i have implemented please check out my fork of graph tag.
> 
> ![IMG_9020](https://private-user-images.githubusercontent.com/3996229/354634391-aa30dca3-3662-4fe9-9c7f-b4c203d1c01d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2MTc1MjAsIm5iZiI6MTcyMjYxNzIyMCwicGF0aCI6Ii8zOTk2MjI5LzM1NDYzNDM5MS1hYTMwZGNhMy0zNjYyLTRmZTktOWM3Zi1iNGMyMDNkMWMwMWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDJUMTY0NzAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWE2MmVlNmFhNWRhYjBhMDJhMmVhMGE5NDFmNzNjNzQ5NjhhMDM3YjhkYmE3Y2ZiNTFlMjA0MDI5NGVjYzVhZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.YxrNzd2aReN2EgHuca68tdNBDrM-Z4KA7ActqM0hgJ0) ![IMG_9021](https://private-user-images.githubusercontent.com/3996229/354634404-e81667d9-402d-46cf-adb8-0eba455291f0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2MTc1MjAsIm5iZiI6MTcyMjYxNzIyMCwicGF0aCI6Ii8zOTk2MjI5LzM1NDYzNDQwNC1lODE2NjdkOS00MDJkLTQ2Y2YtYWRiOC0wZWJhNDU1MjkxZjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDJUMTY0NzAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Yzk0N2JlNTBjMjliMDdlODRkYWNhYThjM2M5YjE2NDY0MTVmNmNhZWU1MDY3NTJmYjRmYTA1MTEzMDNhN2JiNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.hJn0NQB6Hf-IbgaaFttRrJNl8t6CIMstPi7EMNZB06U)

@KylinMountain Thank you for sharing that! I am looking into your commits, it seems like the most relevant update is these get_index_data methods you implemented. Just want to confirm that!
webserver/search/indexdata.py
https://github.com/microsoft/graphrag/commit/14f24b8a1d1ac781b6cf036d76094018c7011416#diff-0993d32b8e1fe7190cf34482a2164a18e677224b487345b9fe465b6346080f56

Also, it will be great if you can share the bootstrap instructions for your fork so that we can try your updates easier!

**Reply by KylinMountain** [reply]
I have updated the README

**Reply by CHENGLING-QIU** [reply]
> I have updated the README

@KylinMountain  Thank you! I was able to host your improved version and some minor changes to redirect the references. It works well and I am just curious is there a chat UI built? Since I believe you have some streaming output optimized

**Reply by KylinMountain** [reply]
you can download the desktop app at here https://cherry-ai.com/

**Comment by lawyinking** [comment]
It seems like we can only trace back to the original text when the citation is 'source' (which actually refers to 'create_base_text_units.parquet').

For answers with citations that are not 'source', but 'relations', 'communities', 'entities', etc., I believe it is impossible to trace back to the original text. These are summarized through procedures during the construction of the graph by the LLM. Ultimately, the traceback would consist of a series of 'text_unit_ids' (that's why it has coherence in RAG, it is a graph constructed by LLM, it answering us using the 'relationship' info) used to summarize those entities, communities, etc., as shown in the example below in the .parquet file:

`['0212e3baf878b6ad7b7e2f4ba289104a', '04b7c2cab0cc1098d8606cec07f7c8d5', '087fb7fb40330a1589f876a93ee83dde', ...]`

to achieve a better accuracy of RAG, best way today should be Graph, but graph has this limitation.

**Reply by lawyinking** [reply]
If you prefer GraphRAG to answer based only on 'source' sometimes, you can type something like `'... answer with reference only to data sources, but not to entities, relationships, or communities'` in your prompt, ofc It's only for local search (if you prefer not 'too global'), and it's meaningless for global search.
